### PR TITLE
[segmentation] Fix the dotproduct calculation in `extractEuclideanClusters` for smooth surfaces

### LINKED ISSUE #3842: [extract_clusters.h] ExtractEuclideanClusters with angles doesn't view densly sampled cylinder as single cluster
# Description
I feel like there are two errors in the extractEuclideanClusters method when it's used with an angle.
First of all the angle is not compared with the normal of the point which was used for the nearestNeighborSearch but simply with the normal of the first point which was used in the current segment.
Secondly i think the norm of the dotproduct should be taken and not the norm of the arccos of the dotproduct. In my mind the angle between two planes can be between 0° and 90° and not between 0° and 180°.
Taking the norm of the arccos doesn't do anything in this context since the arccos of [-1;1] is always positive anyway. Taking the norm of the dotproduct on the other hand does make sense since this makes the solution for the normal calculation unambiguous and leaves resulting angles in [0°,90°].

Concretely I'm talking about it's usage as f.ex. in https://github.com/PointCloudLibrary/pcl/blob/master/examples/segmentation/example_extract_clusters_normals.cpp
Now I am not 100% sure that these are bugs. Maybe I'm simply missing something. Or maybe the algorithm is mend to work that way?!? I have to say though that this approach seems rather unintuitive and unstable to me. But in the case it indeed is mend that way an explanation for this implementation is very welcome. It certainly isn't clear from the description of the method that it's supposed to work that way.

# Original

in  pcl/segmentation/include/pcl/segmentation/extract_clusters.h line 149 - 152:
```
double dot_p = normals.points[i].normal[0] * normals.points[nn_indices[j]].normal[0] +
               normals.points[i].normal[1] * normals.points[nn_indices[j]].normal[1] +
               normals.points[i].normal[2] * normals.points[nn_indices[j]].normal[2];

if ( std::abs (std::acos (dot_p)) < eps_angle )
```

# Proposal
```
double dot_p = normals.points[seed_queue[sq_idx]].normal[0] * normals.points[nn_indices[j]].normal[0] +
               normals.points[seed_queue[sq_idx]].normal[1] * normals.points[nn_indices[j]].normal[1] +
               normals.points[seed_queue[sq_idx]].normal[2] * normals.points[nn_indices[j]].normal[2];

if ( std::acos (std::abs (dot_p)) < eps_angle )
```

# Cylinder Clustering Example
original solution gives:
was_mach_der_regenbogen_hier_O_O
proposed solution gives:
Screenshot_2020-04-01_18-54-33

### PR DESCRIPTION
**Pull request by bene-inmach**
Fixes #3842

Part 2: dotproduct calculation

### CODE REVIEW

**Comment by kunaltyagi** [comment]
Resolve conflicts please

File: segmentation/include/pcl/segmentation/extract_clusters.h [hunk 1]
```diff
@@ -146,9 +146,9 @@ namespace pcl
 
           //processed[nn_indices[j]] = true;
           // [-1;1]
-          double dot_p = normals.points[i].normal[0] * normals.points[nn_indices[j]].normal[0] +
-                         normals.points[i].normal[1] * normals.points[nn_indices[j]].normal[1] +
-                         normals.points[i].normal[2] * normals.points[nn_indices[j]].normal[2];
+          double dot_p = normals.points[seed_queue[sq_idx]].normal[0] * normals.points[nn_indices[j]].normal[0] +
```
Comment (kunaltyagi) [inline segmentation/include/pcl/segmentation/extract_clusters.h:149]: Isn't `normals.points[x]` equivalent to `normals[x]`? We could improve some readability here

**Comment by kunaltyagi** [comment]
@mvieth I could use your help over here 😆 

@bene-inmach Please rebase on master and resolve the merge conflict please

File: segmentation/include/pcl/segmentation/extract_clusters.h [hunk 2]
```diff
@@ -251,10 +251,9 @@ namespace pcl
 
           //processed[nn_indices[j]] = true;
           // [-1;1]
-          double dot_p =
-            normals.points[indices[i]].normal[0] * normals.points[indices[nn_indices[j]]].normal[0] +
-            normals.points[indices[i]].normal[1] * normals.points[indices[nn_indices[j]]].normal[1] +
-            normals.points[indices[i]].normal[2] * normals.points[indices[nn_indices[j]]].normal[2];
+          double dot_p = normals[seed_queue[sq_idx]].normal[0] * normals[nn_indices[j]].normal[0] +
+                         normals[seed_queue[sq_idx]].normal[1] * normals[nn_indices[j]].normal[1] +
+                         normals[seed_queue[sq_idx]].normal[2] * normals[nn_indices[j]].normal[2];
```
Comment (mvieth) [inline segmentation/include/pcl/segmentation/extract_clusters.h:256]: I believe the `indices` has to stay

**Review by mvieth — COMMENTED** [review]
In general, I would not say that the current behaviour is necessarily an error, it simply enforces that all points in a cluster must have similar normals. However, I support this PR since it makes the behaviour normal-wise similar to the behaviour distance-wise: with the distance, it is not necessary that all points in a cluster are close to each other, it is sufficient if two distant points have intermediate points that connect them. With the normals that would mean that two points with very different normals can belong to the same cluster if there are points whose normals lie between them (this is what this PR implements). So clusters would continue around smooth curves (e.g. cylinders, as @bene-inmach showed in the issue), but would stop at sharp edges (e.g. where two planes meet), correct? I think it would be good if the documentation of these two functions would briefly describe this behaviour.
BTW while looking at `extract_cluster.h[pp]` I noticed general room for modernization, e.g. range-for-loops, type-issues, `index_t`, possibly merging of duplicate code, but that can wait for another time/another PR.

**Comment by bene-inmach** [comment]
I resolved the merge conflicts. Sorry for the delay.
Now I did not keep the "indices" as proposed by @mvieth.
It would be great if you could check this again. My reasoning is that the "radiusSearch"-function says in its description about the indices vector: k_indices the resultant indices of the neighboring points.
Therefore I would assume that the resulting vector already points to the indices in the PointCloud directly and not to the indices of a vector containing the indices pointing to the Pointcloud.
Any feedback on this is very welcome.

**Comment by mvieth** [comment]
After thinking about it, I would say you are right - the `indices` has to go and your code is correct like this. I am just surprised that nobody had any problems with it so far.

**Comment by mvieth** [comment]
@bene-inmach Ok, so I confirmed again that there is a bug in the current implementation (master), which causes a segfault (lines 269-271). It vanishes if `normals[indices[nn_indices[j]]]` is changed to `normals[nn_indices[j]]`, like you did in the commit.
If you add a short explanation of the (new) behaviour in the doxygen comments of the two functions (regarding the normals), I am good with merging it.

**Comment by bene-inmach** [comment]
@mvieth I adjusted the doxygen doc only a little bit to make clear that the indices for euclidean distance comparison are the same as for normal angle comparison. It seems to me that this behavior is what is expected from the description much rather than what the function was doing before

**Comment by mvieth** [comment]
Hm, I think that the documentation should at least mention that clusters can extend around smoothly curved surfaces, but stop at sharp edges. It is hard to predict what people expect, it is better to describe things in detail.
BTW one of the comments says that `eps_angle` is in degrees, while the other correctly says it is in radians. If you could also fix that, that would be great.

File: segmentation/include/pcl/segmentation/extract_clusters.h [hunk 3]
```diff
@@ -84,7 +84,10 @@ namespace pcl
 
   //////////////////////////////////////////////////////////////////////////////////////////////////////////////////
   /** \brief Decompose a region of space into clusters based on the euclidean distance between points, and the normal
-    * angular deviation
+    * angular deviation between points. Each point added to the cluster is origin to another radius search. Each point
+    * within radius range will be compared to the origin in respect to normal angle and euclidean distance. If both
+    * are under their respecitve threshhold the point will be added to the cluster. Generally speaking the cluster
```
Comment (mvieth) [inline segmentation/include/pcl/segmentation/extract_clusters.h:89]:
```suggestion
    * are under their respective threshold the point will be added to the cluster. Generally speaking the cluster 
```

File: segmentation/include/pcl/segmentation/extract_clusters.h [hunk 4]
```diff
@@ -188,15 +191,18 @@ namespace pcl
...
```
Comment (mvieth) [inline segmentation/include/pcl/segmentation/extract_clusters.h:196]:
```suggestion
    * are under their respective threshold the point will be added to the cluster. Generally speaking the cluster 
```

**Review by mvieth — APPROVED** [review]
@kunaltyagi Do you want to take a look at the new changes?

**Review by kunaltyagi — APPROVED** [review]

**Pull request merged**